TASK: Add signals to notify on asset add / update / remove

daniellienert opened the pull request:
In order to implement additional tasks for asset handling,
like meta data management, these new signals in the
central point of asset handling - the meta data repository -
notify on asset created / updated / removed.

Moved assetCreated and thumbnailCreated to the corresponding
services for consistency.

aertmann commented:
not sure if this a good way to go tbh.

this is probably better done using a doctrine event listener than adding signals for all these "events", it will quickly fall short when it's needed for another model in the collection..

additionally there is already a ``assetCreated`` event emitted from the asset model itself, see 307cd232fd1cd4a486cbbae1efc75f62a75c099a.. and to be honest even those have weak reasoning for being there, they were added for an easy extension point needed for asynchronous thumbnail generation.

dfeyer commented:
As we plan to move asset to the cr at some point I think it's good to have those signal the api will stay stable when we change the storage

aertmann commented:
@dfeyer: that's a fair point, although I wonder if the asset repository will keep as it is and model the way it is now when switching to CR.. could make sense for backwards compatibility, but also a likely cause of additional problems

also what about the duplicate `assetCreated` events?

daniellienert commented:
> this is probably better done using a doctrine event listener
I also thought about that. But wouldn't that mean to catch ALL doctrine storage events, leading to build the handling objects and execute the code that 99% of the time decides, that it is not asset related. Wouldn't that slow down everything a lot if we add more of these listeners?

> additionally there is already a assetCreated event emitted
Yeah, I could use that, but than I still need the deleted / updated signals. If we would have another central instance to emit asset signals, I would rather emit them there. But ATM, it seems the repository is the only central component.

dfeyer commented:
> Yeah, I could use that, but than I still need the deleted / updated signals. If we would have another central instance to emit asset signals, I would rather emit them there. But ATM, it seems the repository is the only central component.

Why not introducing an AssetSignalService to decouple this from the repository ?

daniellienert commented:
Great idea @dfeyer. I moved the signals to a new AssetSignalService. 

I also moved the emitAssetResourceReplaced signal from the assetService to the assetSignalService and changed the usage in the Neos Package. @radmiraal would that be ok? Do we need to deprecate this signal - it was never released?

@aertmann I would like to deprecate the emitAssetCreated signal in the asset. WDYT? And what about the emitThumbnailCreated in Thumbnail. Should / can we also move that signal?

radmiraal commented:
> Why not introducing an AssetSignalService to decouple this from the repository ?

> I also moved the emitAssetResourceReplaced signal from the assetService to the assetSignalService and changed the usage in the Neos Package

Why wouldn't we just put the signals in the asset service? We don't need an extra object for that right?

radmiraal commented:
> And what does StyleCI actually want me to do?
Add a newline on the end of the file ;-)

daniellienert commented:
> Why wouldn't we just put the signals in the asset service? We don't need an extra object for that right?
Thanks for the feedback @radmiraal - moved the signals to the AssetService.

aertmann commented:
@daniellienert: when deprecating the existing signal, you need to adjust https://github.com/neos/neos-development-collection/blob/master/TYPO3.Media/Classes/TYPO3/Media/Package.php#L29

regarding the `emitThumbnailCreated`, you mean adding similar signals to the `ThumbnailService`? if so, I guess that makes sense

daniellienert commented:
@radmiraal, @aertmann, @dfeyer - moved the signals and adjusted the dispatcher. Any further comments?

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php:
```diff
@@ -310,5 +311,26 @@ public function remove($object)
+
+    /**
+     * @param AssetInterface $object
+     * @throws \TYPO3\Flow\Persistence\Exception\IllegalObjectTypeException
```
two nested exception annotations

daniellienert commented:
Thanks @aertmann, done.

daniellienert commented:
Hey @radmiraal, @aertmann, @dfeyer,
if you don't have any further objections, I would like to merge this one tomorrow.
